File: tests/daemon.test.ts
import { afterEach, describe, expect, it, vi } from "vitest";
import * as fs from "fs";
import * as os from "os";
import * as path from "path";
import { CodebaseDaemon } from "../src/daemon";

type ManagerStub = {
  upsertFileContextNode: ReturnType<typeof vi.fn>;
  deleteContextNode: ReturnType<typeof vi.fn>;
};

function createManagerStub(): ManagerStub {
  return {
    upsertFileContextNode: vi.fn().mockResolvedValue(undefined),
    deleteContextNode: vi.fn().mockResolvedValue(undefined),
  };
}

function makeTempDir(): string {
  return fs.mkdtempSync(path.join(os.tmpdir(), "contextfs-daemon-test-"));
}

function source(lines: string[]): string {
  return lines.join("\n");
}

afterEach(() => {
  vi.restoreAllMocks();
});

describe("CodebaseDaemon", () => {
  it("stores compact logic graph with module-internal symbols and call edges", async () => {
    const tempDir = makeTempDir();
    const nestedDir = path.join(tempDir, "src", "domain");
    fs.mkdirSync(nestedDir, { recursive: true });
    const filePath = path.join(nestedDir, "feature.ts");
    const code = source([
      "import { slugify } from './slug';",
      "const INTERNAL_SEED = 42;",
      "",
      "function normalize(name: string) {",
      "  return slugify(`${name}-${INTERNAL_SEED}`);",
      "}",
      "",
      "export function greet(name: string) {",
      "  return normalize(name);",
      "}",
      "",
      "export class UserService {",
      "  public run(input: string) {",
      "    this.bump();",
      "    return greet(input);",
      "  }",
      "  private bump() {",
      "    return normalize('x');",
      "  }",
      "}",
    ]);
    fs.writeFileSync(filePath, code, "utf8");

    const manager = createManagerStub();
    const daemon = new CodebaseDaemon(manager as any, "proj", tempDir);

    await (daemon as any).processFile(filePath);

    expect(manager.upsertFileContextNode).toHaveBeenCalledTimes(1);
    const call = manager.upsertFileContextNode.mock.calls[0];
    const [uri, name, abstractText, overviewText, content, parentUri, project, metadata] = call;

    expect(uri).toBe("contextfs://proj/src/domain/feature.ts");
    expect(name).toBe("feature.ts");
    expect(parentUri).toBe("contextfs://proj/src/domain");
    expect(project).toBe("proj");
    expect(metadata).toEqual({ type: "file", path: filePath });

    expect(abstractText).toBe("");
    expect(overviewText).toBe("");
    expect(content).toContain("File: src/domain/feature.ts");
    expect(content).toContain("Language: ts");
    expect(content).toContain("LogicGraph: v1");
    expect(content).toContain("Symbols:");
    expect(content).toContain("- fn fn:greet");
    expect(content).toContain("- fn fn:normalize");
    expect(content).toContain("- mtd mtd:UserService.run");
    expect(content).toContain("Edges:");
    expect(content).toContain("- call fn:greet -> fn:normalize");
    expect(content).toContain("- call mtd:UserService.run -> mtd:UserService.bump");
    expect(content).toContain("- call mtd:UserService.run -> fn:greet");
    expect(content).toContain("- import file -> module:./slug");
    expect(content).not.toContain("return slugify");
  });

  it("falls back to default abstract and overview when no declarations exist", async () => {
    const tempDir = makeTempDir();
    const filePath = path.join(tempDir, "helpers.ts");
    const code = source([
      "/* empty module */",
    ]);
    fs.writeFileSync(filePath, code, "utf8");

    const manager = createManagerStub();
    const daemon = new CodebaseDaemon(manager as any, "proj", tempDir);

    await (daemon as any).processFile(filePath);

    expect(manager.upsertFileContextNode).toHaveBeenCalledTimes(1);
    const [, , abstractText, overviewText, content] = manager.upsertFileContextNode.mock.calls[0];

    expect(abstractText).toBe("");
    expect(overviewText).toBe("");
    expect(content).toContain("File: helpers.ts");
    expect(content).toContain("Language: ts");
    expect(content).toContain("Symbols:");
    expect(content).toContain("- (none)");
    expect(content).toContain("Edges:");
    expect(content).not.toContain("empty module");
  });

  it("skips files larger than the configured size limit", async () => {
    const tempDir = makeTempDir();
    const filePath = path.join(tempDir, "large.ts");
    const code = source([`export const payload = "${"x".repeat(2048)}";`]);
    fs.writeFileSync(filePath, code, "utf8");

    const manager = createManagerStub();
    const daemon = new CodebaseDaemon(manager as any, "proj", tempDir, {
      maxFileSizeBytes: 256,
    });

    await (daemon as any).processFile(filePath);

    expect(manager.upsertFileContextNode).not.toHaveBeenCalled();
  });

  it("skips files inside ignored directories", async () => {
    const tempDir = makeTempDir();
    const ignoredFile = path.join(tempDir, "node_modules", "pkg", "index.ts");
    fs.mkdirSync(path.dirname(ignoredFile), { recursive: true });
    fs.writeFileSync(ignoredFile, source(["export const fromDependency = true;"]), "utf8");

    const manager = createManagerStub();
    const daemon = new CodebaseDaemon(manager as any, "proj", tempDir);

    await (daemon as any).processFile(ignoredFile);

    expect(manager.upsertFileContextNode).not.toHaveBeenCalled();
  });

  it("removes deleted files and clears internal caches", async () => {
    const tempDir = makeTempDir();
    const filePath = path.join(tempDir, "module.ts");
    const code = source([
      "export function ping() {",
      "  return 'pong';",
      "}",
    ]);
    fs.writeFileSync(filePath, code, "utf8");

    const manager = createManagerStub();
    const daemon = new CodebaseDaemon(manager as any, "proj", tempDir);

    await (daemon as any).processFile(filePath);
    expect((daemon as any).fileContentHashes.has(path.resolve(filePath))).toBe(true);

    fs.unlinkSync(filePath);
    await (daemon as any).handleFileDelete(filePath);

    expect(manager.deleteContextNode).toHaveBeenCalledTimes(1);
    expect((daemon as any).fileContentHashes.has(path.resolve(filePath))).toBe(false);
    expect((daemon as any).fileFingerprints.has(path.resolve(filePath))).toBe(false);
    expect((daemon as any).nodePayloadHashes.has(path.resolve(filePath))).toBe(false);
  });

  it("skips mtime-only changes when file content is unchanged", async () => {
    const tempDir = makeTempDir();
    const filePath = path.join(tempDir, "module.ts");
    fs.writeFileSync(
      filePath,
      source([
        "export function ping() {",
        "  return 'pong';",
        "}",
      ]),
      "utf8"
    );

    const manager = createManagerStub();
    const daemon = new CodebaseDaemon(manager as any, "proj", tempDir);

    await (daemon as any).processFile(filePath);
    fs.utimesSync(filePath, new Date(Date.now() + 1000), new Date(Date.now() + 1000));
    await (daemon as any).processFile(filePath);

    expect(manager.upsertFileContextNode).toHaveBeenCalledTimes(1);
  });

  it("skips context upsert when compact payload does not change", async () => {
    const tempDir = makeTempDir();
    const filePath = path.join(tempDir, "math.ts");
    fs.writeFileSync(
      filePath,
      source([
        "export function add(a: number, b: number) {",
        "  return a + b;",
        "}",
      ]),
      "utf8"
    );

    const manager = createManagerStub();
    const daemon = new CodebaseDaemon(manager as any, "proj", tempDir);

    await (daemon as any).processFile(filePath);
    fs.writeFileSync(
      filePath,
      source([
        "export function add(a: number, b: number) {",
        "  const sum = a + b;",
        "  return sum;",
        "}",
      ]),
      "utf8"
    );
    await (daemon as any).processFile(filePath);

    expect(manager.upsertFileContextNode).toHaveBeenCalledTimes(1);
  });

  it("enforces compact payload bounds with truncation markers", async () => {
    const tempDir = makeTempDir();
    const filePath = path.join(tempDir, "huge.ts");
    const lines: string[] = [];
    for (let i = 0; i < 180; i += 1) {
      lines.push(`export function f${i}(value: number) { return value + ${i}; }`);
    }
    fs.writeFileSync(filePath, source(lines), "utf8");

    const manager = createManagerStub();
    const daemon = new CodebaseDaemon(manager as any, "proj", tempDir);

    await (daemon as any).processFile(filePath);

    expect(manager.upsertFileContextNode).toHaveBeenCalledTimes(1);
    const [, , , , content, , , metadata] = manager.upsertFileContextNode.mock.calls[0];
    expect(metadata).toEqual({ type: "file", path: filePath });
    expect(content).toContain("GraphStats:");
    expect(content).toContain("Truncated:");
    expect(content.length).toBeLessThanOrEqual(16_100);
  });
});

File: mairu/internal/daemon/daemon_test.go
package daemon

import (
	"context"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"testing"
	"time"
)

type upsertCall struct {
	URI      string
	Name     string
	Abstract string
	Overview string
	Content  string
	Parent   string
	Project  string
	Metadata map[string]any
}

type managerStub struct {
	mu      sync.Mutex
	upserts []upsertCall
	deletes []string
}

func (m *managerStub) UpsertFileContextNode(_ context.Context, uri, name, abstractText, overviewText, content, parentURI, project string, metadata map[string]any) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.upserts = append(m.upserts, upsertCall{
		URI: uri, Name: name, Abstract: abstractText, Overview: overviewText, Content: content, Parent: parentURI, Project: project, Metadata: metadata,
	})
	return nil
}

func (m *managerStub) DeleteContextNode(_ context.Context, uri string) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.deletes = append(m.deletes, uri)
	return nil
}

func TestProcessFileStoresAbstractOverviewAndContent(t *testing.T) {
	dir := t.TempDir()
	nested := filepath.Join(dir, "src", "domain")
	if err := os.MkdirAll(nested, 0o755); err != nil {
		t.Fatal(err)
	}
	file := filepath.Join(nested, "feature.ts")
	src := strings.Join([]string{
		"import { slugify } from './slug';",
		"function normalize(name: string) { return slugify(name); }",
		"export function greet(name: string) { return normalize(name); }",
		"export class UserService { public run(input: string) { return greet(input); } }",
	}, "\n")
	if err := os.WriteFile(file, []byte(src), 0o644); err != nil {
		t.Fatal(err)
	}

	mgr := &managerStub{}
	d := New(mgr, "proj", dir, Options{})
	if err := d.ProcessFile(context.Background(), file); err != nil {
		t.Fatalf("process failed: %v", err)
	}
	if len(mgr.upserts) != 1 {
		t.Fatalf("expected one upsert, got %d", len(mgr.upserts))
	}
	call := mgr.upserts[0]
	if call.URI != "contextfs://proj/src/domain/feature.ts" {
		t.Fatalf("unexpected uri: %s", call.URI)
	}
	if !strings.Contains(call.Overview, "File: src/domain/feature.ts") || !strings.Contains(call.Overview, "Symbols:") {
		t.Fatalf("unexpected overview: %s", call.Overview)
	}
	if call.Abstract == "" || call.Content == "" {
		t.Fatalf("expected non-empty abstract/content")
	}
}

func TestSkipsLargeFilesAndIgnoredDirs(t *testing.T) {
	dir := t.TempDir()
	large := filepath.Join(dir, "large.ts")
	if err := os.WriteFile(large, []byte("export const p = \""+strings.Repeat("x", 2048)+"\";"), 0o644); err != nil {
		t.Fatal(err)
	}
	ignored := filepath.Join(dir, "node_modules", "pkg")
	if err := os.MkdirAll(ignored, 0o755); err != nil {
		t.Fatal(err)
	}
	ignoredFile := filepath.Join(ignored, "index.ts")
	if err := os.WriteFile(ignoredFile, []byte("export const x = 1"), 0o644); err != nil {
		t.Fatal(err)
	}
	mgr := &managerStub{}
	d := New(mgr, "proj", dir, Options{MaxFileSizeBytes: 256})
	_ = d.ProcessFile(context.Background(), large)
	_ = d.ProcessFile(context.Background(), ignoredFile)
	if len(mgr.upserts) != 0 {
		t.Fatalf("expected no upserts, got %d", len(mgr.upserts))
	}
}

func TestDeleteClearsCachesAndCallsManager(t *testing.T) {
	dir := t.TempDir()
	file := filepath.Join(dir, "module.ts")
	_ = os.WriteFile(file, []byte("export function ping(){ return 'pong'; }"), 0o644)
	mgr := &managerStub{}
	d := New(mgr, "proj", dir, Options{})
	_ = d.ProcessFile(context.Background(), file)
	if _, ok := d.fileContentHashes[file]; !ok {
		t.Fatal("expected file hash to be cached")
	}
	if err := os.Remove(file); err != nil {
		t.Fatal(err)
	}
	if err := d.HandleFileDelete(context.Background(), file); err != nil {
		t.Fatal(err)
	}
	if len(mgr.deletes) != 1 {
		t.Fatalf("expected one delete call, got %d", len(mgr.deletes))
	}
	if _, ok := d.fileContentHashes[file]; ok {
		t.Fatal("expected cache cleared")
	}
}

func TestSkipsMtimeOnlyChanges(t *testing.T) {
	dir := t.TempDir()
	file := filepath.Join(dir, "mod.ts")
	_ = os.WriteFile(file, []byte("export function ping(){ return 'pong'; }"), 0o644)
	mgr := &managerStub{}
	d := New(mgr, "proj", dir, Options{})
	_ = d.ProcessFile(context.Background(), file)
	now := time.Now().Add(time.Second)
	_ = os.Chtimes(file, now, now)
	_ = d.ProcessFile(context.Background(), file)
	if len(mgr.upserts) != 1 {
		t.Fatalf("expected one upsert, got %d", len(mgr.upserts))
	}
}

func TestReUpsertsOnContentChange(t *testing.T) {
	dir := t.TempDir()
	file := filepath.Join(dir, "math.ts")
	_ = os.WriteFile(file, []byte("export function add(a:number,b:number){ return a+b; }"), 0o644)
	mgr := &managerStub{}
	d := New(mgr, "proj", dir, Options{})
	_ = d.ProcessFile(context.Background(), file)
	_ = os.WriteFile(file, []byte("export function add(a:number,b:number){ const s=a+b; return s; }"), 0o644)
	time.Sleep(2 * time.Millisecond)
	now := time.Now().Add(2 * time.Second)
	_ = os.Chtimes(file, now, now)
	_ = d.ProcessFile(context.Background(), file)
	if len(mgr.upserts) != 2 {
		t.Fatalf("expected 2 upserts, got %d", len(mgr.upserts))
	}
}
